### ISSUE TITLE
Add executions of scheduled transactions in transaction panels

### ISSUE THREAD

**Issue by georgeef** [moneymanagerex/moneymanagerex#6860]
This is a proposal to add an option to include scheduled transactions in transaction panels.

## Motivation

The transaction panels (`mmCheckingPanel`) currently show only entries in the `CHECKINGACCOUNT_V1` table, while the scheduled transactions panel (`mmBillsDepositsPanel`) shows only entries in the `BILLSDEPOSITS_V1` table. This simplifies the implementation, since each panel needs to deal with only one type of transactions.

From the user's point of view, a single list containing both executed and scheduled transactions, in a selected period from past to future, gives a better overview. It also makes the processing of transactions more convenient, since a single interface can provide the functionality which is curretly separated in two interfaces.

The transaction panels are most suitable for the display and processing of each individual transaction, either executed or scheduled, either in the past or in the future. They should give options to select any desirable combination. On the other hand, the report panels (mostly) present aggregate information, and they should also give options to select any combination between executed and scheduled transactions, or between past and future.

This issue proposes a fusion between (already executed) transactions and future executions of scheduled transactions, in one panel. Another issue will propose a similar fusion in the report panels.

## Implementation

- Define a new class, which represents either a transaction or an execution of a scheduled transaction. A single scheduled transaction unrolls to one or more executions, which differ only in the execution time (`TRANSDATE`). This class contains a member field which represents the repetition number, starting from 1 for executions of a scheduled transaction, and set to 0 for transactions. This field discriminates between the two cases. 
 
- A database typically contains only a limited number of scheduled transactions, however the list of their executions may be large, depending on the frequency and the selected period. In order to improve efficiency, an execution of a scheduled transactions is represented by the execution date, the repetition number, and a link to the original scheduled transaction.

- The transactions and the scheduled transactions are fused together as follows: a list of transactions is generated as before; a list of scheduled transactions is generated, if this options has been selected; a list of executions is generated and sorted by date. The two lists (transactions and executions of scheduled transactions) are then processed in parallel; in each iteration the one with the smaller date (`TRANSDATE`) is selected and stored.

- Once a single list of fused transactions is available, most of the interface functionality is the same for all, however several actions must differentiate between a transaction and an execution of a scheduled transaction.

## Task list
- [X] #6861
- [X] #6873
- [X] #6878
- [X] #6880
- [X] #6981
- [X] #6983
- [X] #7081
- [X] #7084

**Comment by georgeef** [comment]
related issue: #2225

**Comment by whalley** [comment]
@guanlisheng @n-stein Do you think it worthwhile holding up the 1.8.1 release for this and perhaps upping the next release to 1.9.0 to cover the significance of this change?

**Comment by n-stein** [comment]
Agreed, this is pretty significant. I think 1.8.1 should be released before incorporating this. I was working under the assumption 1.8.1 would be released shortly, so I vote we button it up and release in the next few weeks.

**Comment by guanlisheng** [comment]
yup, v1.9.0 is better.

**Comment by whalley** [comment]
> Agreed, this is pretty significant. I think 1.8.1 should be released before incorporating this. I was working under the assumption 1.8.1 would be released shortly, so I vote we button it up and release in the next few weeks.

OK let's get 1.8.1 out the way before merging this work.

**Comment by guanlisheng** [comment]
looking forward to v1.8.1

**Comment by georgeef** [comment]
@whalley @vomikan 
i'll continue pushing PRs for this issue in pipeline (each on top of the previous). when some documentation is needed, i push a new PR, otherwise just a commit on the last PR.

for new issues i'll start a new pipeline, based on the current mainline. i hope the different pipelines won't diverge too much.

**Comment by whalley** [comment]
@georgeef Some great work that is a real benefit to the project. Thank you.

**Comment by whalley** [comment]
@georgeef Can I suggest that for views where the current filter choice extends into the future without limit that we allow the user to supply a 'number of months to extend', maybe 0-12 where 0 would not extend the transactions. This can then allow the 'All Transactions' and the 'View Transaction Report' options to support this feature (perhaps default to 3?) I suspect most users utilise the 'All Transactions' view and having the feature you have introduced allow future scheduled transactions to be displayed would be really useful.

**Comment by whalley** [comment]
> @whalley I have spent some time In the scope of this issue thinking about more convenient and more flexible ways to define a date range, and I decided to postpone it for another issue, since this topic is big and important in itself. I'll send a new issue about it.

I think quicken just allows the user to specify how far ahead scheduled/reminders transactions ar displayed.. This may be the simplest approach and could work irrespective of whether a time-bounded filter is use or not... https://community.quicken.com/discussion/7877823/future-transactions
